### PR TITLE
Fix segfault when input is corrupted

### PR DESCRIPTION
**Pull request by canbal**
Avoid accessing elements of NULL pointers

### CODE REVIEW

**Comment by jeeb** [comment]
While I cannot yet do a technical review of these changes, please squash cosmetics of your own changes into the commit that introduced those parts of code (you can rebase and force-push the branch in order to update the contents of the pull request).

Cheers.

**Comment by canbal** [comment]
Thanks for the tip. I pushed the cosmetic change back into the original commit.

**Comment by dwbuiten** [comment]
I *think* this is correct.

It should perhaps be merged with a title akin to `timeline: Add proper NULL checks` or something.

@VFR-maniac has last word.

File: core/timeline.c [hunk 1]
```diff
@@ -123,6 +123,7 @@ void isom_timeline_destroy( isom_timeline_t *timeline )
     lsmash_remove_entries( timeline->info_list,  NULL );
     lsmash_remove_entries( timeline->bunch_list, NULL );
     lsmash_free( timeline );
+    timeline = NULL;
```
Comment (VFR-maniac) [inline core/timeline.c]: This makes no sense.

Comment (dwbuiten) [inline core/timeline.c]: Oh, I didn't notice the lack of `**`. Yeah.

Comment (canbal) [inline core/timeline.c]: You're right, my bad. Initially I had the NULL assignment outside of this function call, moved it in without thinking much. I will properly handle this and send another change.

File: core/timeline.c [hunk 2]
```diff
@@ -767,7 +770,12 @@ int isom_timeline_construct( lsmash_root_t *root, uint32_t track_ID )
      || !trak->mdia->mdhd
      ||  trak->mdia->mdhd->timescale == 0
      || !trak->mdia->minf
-     || !trak->mdia->minf->stbl )
+     || !trak->mdia->minf->stbl
+     || !trak->mdia->minf->stbl->stco
+     || !trak->mdia->minf->stbl->stsd
+     || !trak->mdia->minf->stbl->stsz
+     || !trak->mdia->minf->dinf
+     || !trak->mdia->minf->dinf->dref )
```
Comment (VFR-maniac) [inline core/timeline.c]: You need not check if both dinf and dref are NULL.
Even if the both are NULL, shall be correctly handled.

File: core/timeline.c [hunk 3]
```diff
@@ -767,7 +770,12 @@ int isom_timeline_construct( lsmash_root_t *root, uint32_t track_ID )
      || !trak->mdia->mdhd
      ||  trak->mdia->mdhd->timescale == 0
      || !trak->mdia->minf
-     || !trak->mdia->minf->stbl )
+     || !trak->mdia->minf->stbl
+     || !trak->mdia->minf->stbl->stco
+     || !trak->mdia->minf->stbl->stsd
+     || !trak->mdia->minf->stbl->stsz
```
Comment (VFR-maniac) [inline core/timeline.c]: You can remove NULL checks for later st**_entry declarations corresponding those boxes like the following.
-lsmash_entry_t *stco_entry = stco && stco->list ? stco->list->head : NULL;
+lsmash_entry_t *stco_entry = stco->list ? stco->list->head : NULL;

FYI, I might add workaround even if those boxes are absent.

File: core/timeline.c (hunk 2 above)
Comment (canbal) [inline core/timeline.c]: Sorry, this is just an artifact of working on release version v2.3.0 and trying to merge my fixes to master.

File: core/timeline.c (hunk 3 above)
Comment (canbal) [inline core/timeline.c]: - If `stco == NULL` then `int large_presentation = stco->large_presentation || lsmash_check_box_type_identical( stco->type, ISOM_BOX_TYPE_CO64 );` causes a segfault.
- If `stsd == NULL` then `isom_sample_entry_t *description = (isom_sample_entry_t *)lsmash_get_entry_data( &stsd->list, stsc_data ? stsc_data->sample_description_index : 1 );` causes a segfault.
- If `stsz == NULL` then `        constant_sample_size = stsz->sample_size;` can cause a segfault.

Am I missing something?

Comment (VFR-maniac) [inline core/timeline.c]: I mean you add these early NULL checks, then later NULL checks of stco, stsd and stsz are unneeded.

Comment (canbal) [inline core/timeline.c]: I see what you mean now. Correct, I will make the changes.

File: core/timeline.c (hunk 1 above)
Comment (canbal) [inline core/timeline.c]: This is again an artifact of working on an earlier version and trying to merge to master. Dropping this line.

**Comment by VFR-maniac** [comment]
4037946
Applied. Thanks

**Pull request closed without merging**